util/hot_backup.cc

```diff
@@ -0,0 +1,654 @@
+            //   ScanDB() makes calls ... causing trigger to reset too
+            //   early)
+            gHotBackup->HotBackupScheduled();
+            gHotBackup->HotBackupScheduled();
+
+            // Log() routes to syslog as an error
+            Log(NULL, "leveldb HotBackup triggered.");
+            gPerfCounters->Inc(ePerfBackupStarted);
+
+            DBList()->ScanDBs(true,  &DBImpl::HotBackup);
+            DBList()->ScanDBs(false, &DBImpl::HotBackup);
+
+            // reduce to one instance count now that events posted
+            //  (trigger reset now enabled)
+            gHotBackup->HotBackupFinished();
+        }   // if
+    }   // if
+
+    return;
+
+}   // CheckHotBackupTrigger
+
+
+/**
+ * Look for external trigger
+ */
+bool
+HotBackup::IsTriggerSet()
+{
+    bool ret_flag;
+    int ret_val;
+
+    //
+    // This is code polls for existance of /etc/riak/hot_backup
+    //
+    ret_val=access(GetTriggerPath(), F_OK);
+    ret_flag=(-1!=ret_val);
+
+    return(ret_flag);
+
+}   // HotBackup::IsTriggerSet
+
+
+/**
+ * Let external trigger know that hot backup is complete.  Put
+ *  message to syslog if unable to delete trigger file
+ */
+void
+HotBackup::ResetTrigger()
+{
+    bool ret_flag;
+    int ret_val;
+
+    ret_val=unlink(GetTriggerPath());
+    if (0==ret_val)
+    {
+        // release last counter, enabling future backups
+        HotBackupFinished();
+    }   // if
+    else
+    {
+        // one count left, effectively disabling trigger forever (or next program restart)
+        Log(NULL, "leveldb HotBackup unable to delete trigger file %s", GetTriggerPath());
+        Log(NULL, "leveldb HotBackup now disabled until program restart");
+    }   // if
+
+    return;
+
+}   // ResetHotBackupTrigger
+
+
+/**
+ * A database instance marks its intent to backup
+ */
+void
+HotBackup::HotBackupScheduled()
+{
+    inc_and_fetch(&m_JobsPending);
+
+    return;
+
+}   // HotBackupScheduled
+
+
+/**
+ * A database instance marks that its backup completed
+ */
+void
+HotBackup::HotBackupFinished()
+{
+    int ret_val;
+
+    ret_val=dec_and_fetch(&m_JobsPending);
+
+    // 1 means CheckHotBackupTrigger()'s counter is only
+    //   one left
+    if (1==ret_val)
+    {
+        // hmm, call file sync() at this point?
+        Log(NULL, "leveldb HotBackup complete.");
+        ResetTrigger();
+    }   // if
+
+    return;
+
+}   // HotBackupFinished
+
+
+/**
+ * Routine called by DBList()->ScanDBs.  This has full context of one open
+ *  leveldb database.  Validates then schedules a backup job via compaction
+ *  threads.
+ */
+void
+DBImpl::HotBackup()
+{
+    bool create_backup_event(false);
+
+    // hold mutex while reviewing database state
+    //  (note:  never good to call Log() while holding mutex
+    //   but these are unlikely error cases)
+    {
+        // get mutex,
+        MutexLock l(&mutex_);
+
+        // set backup flag && see if
+        assert(false==hotbackup_pending_);
```

Comment:
Ah, so the copy of leveldb embedded in eleveldb still does the correct
thing with asserts? That's good news. I thought we lost all "no asserts in
non-debug builds" behavior when building eleveldb and its dependencies.

On Sun, Sep 4, 2016 at 2:59 PM, Matthew Von-Maszewski <
notifications@github.com> wrote:

> In util/hot_backup.cc
> <https://github.com/basho/leveldb/pull/202#discussion_r77457697>:
>
> > + *  threads.
> > + */
> > +void
> > +DBImpl::HotBackup()
> > +{
> > +    bool create_backup_event(false);
> > +
> > +    // hold mutex while reviewing database state
> > +    //  (note:  never good to call Log() while holding mutex
> > +    //   but these are unlikely error cases)
> > +    {
> > +        // get mutex,
> > +        MutexLock l(&mutex_);
> > +
> > +        // set backup flag && see if
> > +        assert(false==hotbackup_pending_);
>
> Good eye! Wrong repository. Google's Makefile knows to disable assert() in
> production builds and leave active in debug builds. Their asserts have
> saved my ass a couple of times, especially related to keys. (Yes, I
> reworked their Makefile but left that important feature!)
>
> Yes, eleveldb dies on the one assert() I left in that code. But lately it
> has proved better to have that assert() fire than wait for a random, likely
> double free, bug. That said, I do not recommend leaving live asserts as a
> defense mechanism. Someday I will rewrite how the C++ code builds in
> eleveldb.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/basho/leveldb/pull/202/files/9b58e117f4249df583516a6ca3539826525b4b93#r77457697>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMnddPv_Bp65cv4M6gTV2aLF1ge9MLHzks5qmyM7gaJpZM4JcAvk>
> .
>